File: link.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

dotfiles_dir = Dir.pwd
dotfiles = %w[
  doom.d
  git_commit_msg.txt
  gitconfig
  gitignore_global
  zshrc.local
  curlrc
]

dotfiles.each do |file|
  linking_file = "#{Dir.home}/.#{file}"
  linked_file = "#{dotfiles_dir}/#{file}"

  if File.exist?(linking_file)
    File.delete linking_file
    puts "Remove existing file: #{linking_file}"
  end

  File.symlink linked_file, linking_file
  puts "Created symlink for .#{file}", ''
end

File: link.clj
(ns link
  (:require [clojure.java.io :as io])
  (:import [java.nio.file Files] [java.nio.file.attribute FileAttribute]))

(def dotfiles-dir (System/getProperty "user.dir"))
(def user-home (System/getProperty "user.home"))

(def dotfiles
  ["doom.d" "git_commit_msg.txt" "gitconfig"
   "curlrc" "zshrc.local" "gitignore_global"])

(defn- remove-existed [file]
  (when (.exists file)
    (let [path (.getPath file)]
      (println (str "Removing file " path))
      (clojure.java.io/delete-file path))))

(defn- remove-and-link [file]
  (let [linking-file (io/file user-home (str "." file))
       linked-file (io/file dotfiles-dir file)]
    (remove-existed linking-file)
    (Files/createSymbolicLink
     (.toPath linking-file)
     (.toPath linked-file)
     (into-array FileAttribute []))))

(->> dotfiles (map remove-and-link) (map str) println)
